### ISSUE TITLE
[V2] New 'Moderate' vulnerabilities in gatsby dependencies.

### ISSUE THREAD

**Issue by mishasaggi** [gatsbyjs/gatsby#7293]
I see in issue #5335 package vulnerabilities were addressed, however I see some 'Moderate' ones now so these may be new or only for V2.
version: gatsby@2.0.0-beta.99

```
│                                Manual Review                                 │
│            Some vulnerabilities require your attention to resolve            │
│                                                                              │
│         Visit https://go.npm.me/audit-guide for additional guidance          │
└──────────────────────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-pngquant > pngquant-bin > bin-build > download >    │
│               │ caw > tunnel-agent                                           │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-pngquant > pngquant-bin > bin-wrapper > download >  │
│               │ caw > tunnel-agent                                           │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-webp > cwebp-bin > bin-build > download > caw >     │
│               │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-webp > cwebp-bin > bin-wrapper > download > caw >   │
│               │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598     


```

**Comment by mishasaggi** [comment]
Moderate ones:
```
┌──────────────────────────────────────────────────────────────────────────────┐
│                                Manual Review                                 │
│            Some vulnerabilities require your attention to resolve            │
│                                                                              │
│         Visit https://go.npm.me/audit-guide for additional guidance          │
└──────────────────────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-pngquant > pngquant-bin > bin-build > download >    │
│               │ caw > tunnel-agent                                           │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-pngquant > pngquant-bin > bin-wrapper > download >  │
│               │ caw > tunnel-agent                                           │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-webp > cwebp-bin > bin-build > download > caw >     │
│               │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-webp > cwebp-bin > bin-wrapper > download > caw >   │
│               │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598 
```

**Comment by sarahannnicholson** [comment]
I am seeing this as well, I thought I'd add to the details. 

version:  gatsby@2.0.0-rc.21

When I run the command `npm list tunnel-agent` I get the following:
<pre> 
my_project@1.0.0 /mnt/d/MyProject
├─┬ gatsby@2.0.0-rc.21
│ └─┬ request@2.88.0
│   └── tunnel-agent@0.6.0
├─┬ gatsby-plugin-manifest@2.0.2-rc.1
│ └─┬ sharp@0.20.8
│   ├─┬ prebuild-install@4.0.0
│   │ └── tunnel-agent@0.6.0  deduped
│   └── tunnel-agent@0.6.0  deduped
├─┬ <b>gatsby-plugin-sharp@2.0.0-rc.7</b>
│ ├─┬ imagemin-pngquant@6.0.0
│ │ └─┬ pngquant-bin@5.0.0
│ │   ├─┬ bin-build@3.0.0
│ │   │ └─┬ download@6.2.5
│ │   │   └─┬ caw@2.0.1
│ │   │     └── tunnel-agent@0.6.0  deduped
│ │   └─┬ bin-wrapper@3.0.2
│ │     └─┬ download@4.4.3
│ │       └─┬ caw@1.2.0
│ │         └── <b>tunnel-agent@0.4.3</b>
│ └─┬ imagemin-webp@4.1.0
│   └─┬ cwebp-bin@4.0.0
│     └─┬ bin-build@2.2.0
│       └─┬ download@4.4.3
│         └─┬ caw@1.2.0
│           └── <b>tunnel-agent@0.4.3</b>
└─┬ node-sass@4.9.3
  └─┬ request@2.87.0
    └── tunnel-agent@0.6.0  deduped
</pre>

Seems like a dependency of gatsby-plugin-sharp

**Comment by millette** [comment]
I attempted to address that in #7740 but I think it needs to be fixed upstream too.

**Comment by ibanlopez** [comment]
I get same issue but with gatsby-plugin-sharp. Seems to be tunnel-agent:

`├─┬ gatsby@2.0.19
│ └─┬ request@2.88.0
│   └── tunnel-agent@0.6.0 
├─┬ gatsby-plugin-manifest@2.0.5
│ └─┬ sharp@0.20.8
│   ├─┬ prebuild-install@4.0.0
│   │ └── tunnel-agent@0.6.0  deduped
│   └── tunnel-agent@0.6.0  deduped
└─┬ gatsby-plugin-sharp@2.0.7
  ├─┬ imagemin-mozjpeg@7.0.0
  │ └─┬ mozjpeg@5.0.0
  │   └─┬ bin-build@2.2.0
  │     └─┬ download@4.4.3
  │       └─┬ caw@1.2.0
  │         └── **tunnel-agent@0.4.3** 
  └─┬ imagemin-pngquant@6.0.0
    └─┬ pngquant-bin@5.0.0
      └─┬ bin-build@3.0.0
        └─┬ download@6.2.5
          └─┬ caw@2.0.1
            └── tunnel-agent@0.6.0  deduped`

Just because of this: https://github.com/kevva/caw/issues/22

**Comment by SalomonSmeke** [comment]
Hey. Ive been tracking this tunnel-agent stuff for a minute and its not very hopeful, highly recommend forking CAW and shrink-wrapping, or shrink wrapping tunnel-agent to the latest commit hash.

**Comment by jfgrissom** [comment]
This looks like it can be addressed if gatsby-plugin-sharp uses imagemin-mozjpeg@8.0.0 (released about 18 days ago).
It uses new dependencies (mozjpeg@6.0.0) wich in turn uses these.
```
bin-build@3.0.0
│ │ └─┬ download@6.2.5
│ │   └─┬ caw@2.0.1
│ │     └── tunnel-agent@0.6.0 
```

Tunnel-agent@6.0.0 no longer has the vulnerability issue.

**Comment by grantmoran** [comment]
Interested in how to possibly resolve this issue as well.

```
                       === npm audit security report ===

┌──────────────────────────────────────────────────────────────────────────────┐
│                                Manual Review                                 │
│            Some vulnerabilities require your attention to resolve            │
│                                                                              │
│         Visit https://go.npm.me/audit-guide for additional guidance          │
└──────────────────────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-plugin-sharp                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-plugin-sharp > imagemin-mozjpeg > mozjpeg > bin-build │
│               │ > download > caw > tunnel-agent                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-plugin-sharp                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-plugin-sharp > imagemin-mozjpeg > mozjpeg >           │
│               │ bin-wrapper > download > caw > tunnel-agent                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-plugin-sharp                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-plugin-sharp > imagemin-webp > cwebp-bin > bin-build  │
│               │ > download > caw > tunnel-agent                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-plugin-sharp                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-plugin-sharp > imagemin-webp > cwebp-bin >            │
│               │ bin-wrapper > download > caw > tunnel-agent                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-mozjpeg > mozjpeg > bin-build > download > caw >    │
│               │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-mozjpeg > mozjpeg > bin-wrapper > download > caw >  │
│               │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-webp > cwebp-bin > bin-build > download > caw >     │
│               │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-source-contentful                                     │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-source-contentful > gatsby-plugin-sharp >             │
│               │ imagemin-webp > cwebp-bin > bin-wrapper > download > caw >   │
│               │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
found 8 moderate severity vulnerabilities in 36662 scanned packages
  8 vulnerabilities require manual review. See the full report for details.
```

**Comment by daberg2** [comment]
Same issues here. Any word on resolving this issue?
```
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
┌───────────────┬──────────────────────────────────────────────────────────────┐
│ Moderate      │ Memory Exposure                                              │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Package       │ tunnel-agent                                                 │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Patched in    │ >=0.6.0                                                      │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Dependency of │ gatsby-plugin-sharp                                          │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ Path          │ gatsby-plugin-sharp > imagemin-webp > cwebp-bin >            │
│               │ bin-wrapper > download > caw > tunnel-agent                  │
├───────────────┼──────────────────────────────────────────────────────────────┤
│ More info     │ https://nodesecurity.io/advisories/598                       │
└───────────────┴──────────────────────────────────────────────────────────────┘
```

**Comment by jonniebigodes** [comment]
@daberg2 thanks for bringing that to the attention of the team. But as we know this ecosystem is a fast paced, constantly evolving and mutating.
I'm sure that the team is aware of this. And is making plans to update the peer dependencies used by gatsby and the plugins under the care of the team and with that resolve this issue. And if you don't mind  being a bit more patient and be on the lookout for feedback on this issue.

**Comment by ibanlopez** [comment]
@jonniebigodes thanks for mentioning "patience" to the attention of the users that are following this issue since 13 Aug.
I'm sure we, the users, are being patient (me included). Be sure we don't mind to be on the lookout for feedback on this issue and to be very grateful on the team and everything they are doing with Gatsby, because it's awesome. And if you don't mind we will still be pointing out anything new to the team that we see about this issue just to help. So be patient, because we don't know if the team is whether or not aware about this or if they are making plans to update the peer dependencies used by Gatsby and the plugins.
We just see same issue every time we update and we are just trying to help, but thanks to inform us about the awareness and plans of the team.

**Comment by jgierer12** [comment]
Hi @ibanlopez. No need for the toxicity - after all, this is an open source project. Most of us are doing this in our free time. If you feel like this should be fixed: clone the repo, fix it, and submit a PR.

**Comment by jgierer12** [comment]
omg, I'm so sorry @ibanlopez, I interpreted your comment in a passive-aggressive/sarcastic way. (The first sentence in particular, could easily come over as that).

Now that I reread it I can totally see how you meant it. Thanks for clarifying, and sorry again for attacking you so quickly. I hope you don't take it personal :sweat_smile: .

**Comment by ibanlopez** [comment]
@jgierer12 Never mind. I hope I could help to contribute in the project but I'm afraid I don't have the knowledge nor the ability. I'm very thankful for everything all contributors do, it's awesome. I'm sure everybody here are very thankful.
Sorry for the first sentence, I couldn't resist myself after last sentence from jonniebegodes. I just meant that we don't want to rush or piss off anyone, we just try to help the way we can.

**Comment by jonniebigodes** [comment]
i'm refraining from further comments, in order to avoid any more situations like the ones above, i meant what i've said, in the context that i've said it. @ibanlopez sorry if my comment caused any discomfort to you in any way shape or form. But i ask you that in the future be more mindfull of your comments if you don't mind. And i hope you get the chance to further your knowledges and abilities so that sometime in the near future i and other people in here see some contents, plugins, pull requests coming from you. Thanks

**Comment by SalomonSmeke** [comment]
The root of this issue is the usage of the tunnel-agent dependency.

Tunnel-agent contains the vulnerability in all its recent releases. The latest commit no longer has it, but the maintainer of tunnel-agent didn’t npm publish. Thus the issue persists.

In the CAW repository I have made a PR which replaces tunnel-agent with what seems to be the original implementation. Node-tunnel.

One possible solution for this repository is to fork my fork of CAW and use that repository as the source of the CAW dependency in a npm shrinkwrap.

Another is for the CAW maintainer to merge my PR.

Yet another is to point all tunnel-agent dependencies in a shrinkwrap to the latest commit hash from the tunnel-agent repository.

Please let me know if I can help in any way, I check my notifications daily.

Cheers,
Salomon

**Comment by sidharthachatterjee** [comment]
This is now resolved with `gatsby-plugin-sharp@2.0.16`

**Comment by daberg2** [comment]
I can confirm that this issue is resolved now in my setup with `gatsby-plugin-sharp@2.0.16`. Thank you for the fix!
